content/post/flinksql-time.adoc

```diff
@@ -0,0 +1,1033 @@
+We don't want to use a **processing time attribute** because that doesn't answer our question; we need to use an **event time attribute**.
+
+To mark a column as an **event time attribute** we need to use the `WATERMARK` statement.
+This is where I think things get a bit confusing until you understand it, and then it's just… _shrugs_ how Flink is.
+Let me explain…
+
+== Watermarks
+
+When you run a _batch_ query the engine doing the processing knows when it's read all of the data.
+Life is simple.
+Contrast that to a streaming query in which, by definition, the source of the data is unbounded—so there's no such thing as having "read all the data".
+Not only is the source unbounded, but the data may arrive out of order.
+
+**TODO: PUT A DIAGRAM HERE**
+
+_Watermarks_ are a clever idea that tell the processing engine when it's OK to consider a passage of time as complete.
+**Where we set the watermark is up to us**.
+Set a watermark too short and whilst you'll get your final result quicker you're much more likely to have incomplete data because anything arriving late will be ignored.
+Then again, set the watermark too long you'll increase the chances of getting a complete set of data, but at the expense of the result taking longer to finalise.
+
+Which is right? That depends on you and your business process :)
+
+[TIP]
+====
+To learn more about watermarks in detail check out these excellent resources:
+
+* https://www.youtube.com/watch?v=sdhwpUAjqaI[Event Time and Watermarks] (video)
+* https://www.oreilly.com/radar/the-world-beyond-batch-streaming-101/[Streaming 101: The world beyond batch—Tyler Akidau]
+* https://current.confluent.io/2024-sessions/timing-is-everything-understanding-event-time-processing-in-flink-sql[Timing is Everything: Understanding Event-Time Processing in Flink SQL—Sharon Xie]
+====
+
+So watermarks are a thing, and we need to configure them.
+We can't just define a column as an **event time attribute** on its own; we need to define the watermark that goes with it so that Flink knows when to have considered all data as having been read for a given time.
+
+In my mind here is some pseudo-SQL that I'd like to run when working with an event time attribute, but *is not correct Flink SQL*:
+
+[source,sql]
+----
+ALTER TABLE orders_kafka
+    ALTER COLUMN `created_at` TIMESTAMP_LTZ(3) AS EVENT_TIME;
+----
+
+or something like that.
+The point being, **we never explicitly say `this field is the event time attribute`**.
+What we actually do is **_implicitly_ mark it as the event time attribute by defining the watermark**.
+Since there's a watermark, the column on which the watermark is defined must be the event time.
+Obvious, right?!
+
+[source,sql]
+----
+ALTER TABLE orders_kafka
+    ADD WATERMARK FOR `created_at` AS `created_at` - INTERVAL '5' SECOND;
+----
+
+This _basically_ tells Flink that it needs to give a five-second leeway when processing `created_at` for any late records to arrive on the stream.
```

Comment:
From a technical standpoint, Flink considers records late if they are behind the watermark. The five-second leeway is accommodating _out-of-order_ records, and preventing them from being considered late.